java-client/session-dagger/src/test/java/io/deephaven/client/MultiClientSharedIdTest.java

```diff
@@ -54,4 +60,54 @@ public void testHandoff() throws TableHandle.TableHandleException, InterruptedEx
         // Ensure that we can resolve from query scope.
         client2.execute(destId.ticketId().table());
     }
+
+    @Test
+    public void testFailsPublishingFromQueryScope() throws Exception {
+        final TimeTable sourceTable = TimeTable.of(Duration.ofSeconds(1));
+        final TableHandle remoteSource = session.execute(sourceTable);
+        final ScopeId scopeId = new ScopeId("test_time_table");
+        final SharedId sharedId = SharedId.newRandom();
+
+        // Let's publish the source table to the query scope.
+        session.publish(scopeId, remoteSource).get();
+
+        // This publish should fail because the source is not a session owned export.
+        try {
+            session.publish(sharedId, scopeId).get();
+            Assert.statementNeverExecuted();
+        } catch (final ExecutionException err) {
+            // This is expected.
+            Assert.eqTrue(err.getCause() instanceof StatusRuntimeException,
+                    "err.getCause() instanceof StatusRuntimeException");
+        }
+    }
+
+    @Test
+    public void testSharePublishCompletesImmediately() throws Exception {
+        final TicketId ticketId = new TicketId(ExportTicketHelper.exportIdToBytes(1));
+        final SharedId sharedId = SharedId.newRandom();
+
+        // Let's publish the source table to the shared id.
+        session.publish(sharedId, ticketId).get();
+    }
+
+    @Test
+    public void testScopePublishWaitsForSourceCompletion() throws Exception {
+        final TicketId ticketId = new TicketId(ExportTicketHelper.exportIdToBytes(1));
+        final ScopeId scopeId = new ScopeId("test_dest_table");
+
+        // Let's publish the source table to the shared id.
+        final CompletableFuture<Void> rpc = session.publish(scopeId, ticketId);
+        try {
+            rpc.get(250, TimeUnit.MILLISECONDS);
+        } catch (final TimeoutException ignored) {
+            // This is expected.
+        }
+
+        final TimeTable sourceTable = TimeTable.of(Duration.ofSeconds(1));
+        final TableHandle remoteSource = session.execute(sourceTable);
```

Comment:
We are making an _assumption_ here that it will get executed with export id 1 / `ticketId`. It might be better to explicitly control the server side with `io.deephaven.server.runner.DeephavenApiServerSingleAuthenticatedBase#authenticatedSessionState`. If necessary, I think we can have a new method like `io.deephaven.server.session.SessionState#newServerSideExport` where the test framework is expected to call `setResult` if we want to test delayed completion.